### PR TITLE
Allow pre-parsing of ESI document, add example of fetching per-user fragment variants

### LINKED ISSUE #3: Ability to pre-parse document for fragment URLs
For the sake of setting surrogate keys on responses, it is useful to know which fragments are going to be requested before we start executing.

This is not possible for recursive fragments, but we could still parse the document initially to fetch all of the top-level fragments.

### PR DESCRIPTION
**Pull request by kailan**
Resolves #3

### CODE REVIEW

File: esi/src/lib.rs [hunk 1]
```diff
@@ -140,9 +169,8 @@ impl Processor {
         );
 
         // `root_task` is the root task that will be used to fetch tags in recursive manner
-        let root_task = &mut Task::new();
+        let root_task: &mut Task = &mut Task::new();
```
Comment (vagetman) [inline esi/src/lib.rs:172]: I think this is pretty obvious and no more explicit notation required.

File: examples/esi_example_variants/fastly.toml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# This file describes a Fastly Compute package. To learn more visit:
+# https://developer.fastly.com/reference/fastly-toml/
+
+authors = ["kailan@enviark.com"]
```
Comment (vagetman) [inline examples/esi_example_variants/fastly.toml:4]:
```suggestion
authors = ["kblanks@fastly.com"]
```
I think we usually leaving those empty or with fastly emails, no?

File: examples/esi_example_variants/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    io::Write,
+};
```
Comment (vagetman) [inline examples/esi_example_variants/src/main.rs:4]: No need to use `VecDeque`
```suggestion
use std::{collections::HashMap, io::Write};
};
```

File: examples/esi_example_variants/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        let resp = Response::from_status(StatusCode::OK).with_content_type(mime::TEXT_HTML);
+
+        // Send the response headers to the client and open an output stream
+        let output_writer = resp.stream_to_client();
+
+        // Set up an XML writer to write directly to the client output stream.
+        let mut xml_writer = Writer::new(output_writer);
+
+        // Parse the ESI document and store it in memory
+        let mut events = VecDeque::new();
```
Comment (vagetman) [inline examples/esi_example_variants/src/main.rs:59]: No need to use `VecDeque`
```suggestion
        let mut events = Vec::new();
```

File: examples/esi_example_variants/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        let mut beresp_reader = Reader::from_reader(beresp.take_body());
+        parse_tags("esi", &mut beresp_reader, &mut |event| {
+            events.push_back(event);
```
Comment (vagetman) [inline examples/esi_example_variants/src/main.rs:62]: No need to use `VecDeque`
```suggestion
            events.push(event);
```

File: examples/esi_example_variants/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+
+        // Check the variant database to determine the URLs to fetch
+        let variant_urls = get_variant_urls(urls);
+
+        // Process the already-parsed ESI document, replacing the request URLs with the variant URLs
+        let result = processor.process_parsed_document(
+            events,
```
Comment (vagetman) [inline examples/esi_example_variants/src/main.rs:81]: No need to use `VecDeque`
```suggestion
            events.into(),
```

**Review by vagetman — COMMENTED** [review]
Some minor details

**Comment by kailan** [comment]
Thanks for the review @vagetman, I addressed all of those comments. How's it looking now?

**Review by vagetman — APPROVED** [review]

**Pull request merged**